ul/src/pdu/mod.rs

```diff
@@ -0,0 +1,270 @@
+    Reserved,
+}
+
+impl AbortRQSource {
+    fn from(source: u8, reason: u8) -> Option<AbortRQSource> {
+        let result = match source {
+            0 => AbortRQSource::ServiceUser,
+            1 => AbortRQSource::Reserved,
+            2 => match reason {
+                0 => AbortRQSource::ServiceProvider(
+                    AbortRQServiceProviderReason::ReasonNotSpecifiedUnrecognizedPDU,
+                ),
+                2 => AbortRQSource::ServiceProvider(AbortRQServiceProviderReason::UnexpectedPDU),
+                3 => AbortRQSource::ServiceProvider(AbortRQServiceProviderReason::Reserved),
+                4 => AbortRQSource::ServiceProvider(
+                    AbortRQServiceProviderReason::UnrecognizedPDUParameter,
+                ),
+                5 => AbortRQSource::ServiceProvider(
+                    AbortRQServiceProviderReason::UnexpectedPDUParameter,
+                ),
+                6 => AbortRQSource::ServiceProvider(
+                    AbortRQServiceProviderReason::InvalidPDUParameter,
+                ),
+                _ => {
+                    return None;
+                }
+            },
+            _ => {
+                return None;
+            }
+        };
+
+        Some(result)
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, PartialOrd, Hash, Debug)]
+pub enum AbortRQServiceProviderReason {
+    ReasonNotSpecifiedUnrecognizedPDU,
+    UnexpectedPDU,
+    Reserved,
+    UnrecognizedPDUParameter,
+    UnexpectedPDUParameter,
+    InvalidPDUParameter,
+}
+
+#[derive(Clone, Eq, PartialEq, PartialOrd, Hash, Debug)]
+pub enum PduVariableItem {
+    Unknown(u8),
+    ApplicationContext(String),
+    PresentationContextProposed(PresentationContextProposed),
+    PresentationContextResult(PresentationContextResult),
+    UserVariables(Vec<UserVariableItem>),
+}
+
+#[derive(Clone, Eq, PartialEq, PartialOrd, Hash, Debug)]
+pub enum UserVariableItem {
+    Unknown(u8, Vec<u8>),
+    MaxLength(u32),
+    ImplementationClassUID(String),
+    ImplementationVersionName(String),
+}
+
+#[derive(Clone, Eq, PartialEq, PartialOrd, Hash, Debug)]
+pub enum PDU {
+    Unknown {
+        pdu_type: u8,
+        data: Vec<u8>,
+    },
+    AssociationRQ {
+        protocol_version: u16,
+        calling_ae_title: String,
+        called_ae_title: String,
+        application_context_name: String,
+        presentation_contexts: Vec<PresentationContextProposed>,
+        user_variables: Vec<UserVariableItem>,
+    },
+    AssociationAC {
+        protocol_version: u16,
+        application_context_name: String,
+        presentation_contexts: Vec<PresentationContextResult>,
+        user_variables: Vec<UserVariableItem>,
+    },
+    AssociationRJ {
+        result: AssociationRJResult,
+        source: AssociationRJSource,
+    },
+    PData {
+        data: Vec<PDataValue>,
+    },
+    ReleaseRQ,
+    ReleaseRP,
+    AbortRQ {
+        source: AbortRQSource,
+    },
+}
+
+pub mod reader;
```

Comment:
I usually declare modules right before the `use` statements a the top, rather than at the bottom.